File: src/Flare.Api/Model/HostStatsSnapshot.cs
using MemoryPack;

namespace Flare.Api.Model;

/// <summary>
/// The full payload behind both <c>GET /api/resources/host/snapshot</c> and every message
/// pushed over <c>GET /api/resources/host/watch</c> - the Resources page's "Host overview"
/// panel. See <c>HostStats.HostStatsPoller</c> for how it's sampled and broadcast.
/// </summary>
[MemoryPackable]
public sealed partial record HostStatsSnapshot
{
    /// <summary>
    /// <see langword="false"/> means this host's stats genuinely can't be read - not Linux,
    /// or <c>/proc/stat</c> missing (see <c>HostStats.HostStatsPoller</c>'s startup check) -
    /// which happens on a non-Linux Aspire dev loop (e.g. a Mac running <c>Flare.Api</c>
    /// directly). Unlike <see cref="ResourceGraphSnapshot.Available"/> there's no separate
    /// "configured but unreachable" state to distinguish - this either can read
    /// <c>/proc</c> or it can't.
    /// </summary>
    public required bool Available { get; init; }

    /// <summary>Human-readable explanation, set whenever <see cref="Available"/> is <see langword="false"/> - <see langword="null"/> once real data is showing.</summary>
    public string? UnavailableReason { get; init; }

    /// <summary>0-100. 0.0 on the very first sample (no prior <c>/proc/stat</c> reading to diff against yet) - corrects itself on the next poll tick.</summary>
    public double CpuUsagePercent { get; init; }

    /// <summary>True host core count, from counting <c>/proc/stat</c>'s <c>cpuN</c> lines - see <c>HostStats.ProcFileReader.ParseStat</c>'s remarks for why this isn't <see cref="Environment.ProcessorCount"/>.</summary>
    public int CpuCoreCount { get; init; }

    /// <summary>1-minute load average from <c>/proc/loadavg</c>.</summary>
    public double LoadAverage1m { get; init; }

    /// <summary>Per-core usage percent, same 0-100/delta-based shape as <see cref="CpuUsagePercent"/> - index <c>i</c> corresponds to <c>cpuN</c> where <c>N == i</c>. Empty on the very first sample (no prior per-core tick to diff against), same as <see cref="CpuUsagePercent"/> starting at 0.0. Feeds the Host overview panel's per-core breakdown, not the Resource trends chart - see that panel's own remarks on why per-core stays instantaneous-only.</summary>
    public IReadOnlyList<double> PerCoreUsagePercent { get; init; } = [];

    public long MemoryTotalBytes { get; init; }

    public long MemoryUsedBytes { get; init; }

    /// <summary><c>MemAvailable</c> from <c>/proc/meminfo</c> - not simply <see cref="MemoryTotalBytes"/> minus <see cref="MemoryUsedBytes"/> restated (it *is* that value, by construction - <c>MemoryUsedBytes</c> is derived from it - but shown as its own field so the frontend never has to re-derive it, and so the two stay obviously consistent).</summary>
    public long MemoryAvailableBytes { get; init; }

    /// <summary>0 on a host with swap disabled entirely - common under Docker Desktop and various cloud VM images. The frontend hides the swap line whenever this is 0, rather than showing "0 B swap" everywhere.</summary>
    public long SwapTotalBytes { get; init; }

    public long SwapUsedBytes { get; init; }

    public long DiskTotalBytes { get; init; }

    public long DiskUsedBytes { get; init; }

    /// <summary><see cref="DiskTotalBytes"/> minus <see cref="DiskUsedBytes"/> - see <see cref="MemoryAvailableBytes"/>'s remarks on why this is its own field rather than left to the frontend to subtract.</summary>
    public long DiskAvailableBytes { get; init; }

    /// <summary>
    /// Bytes/sec, from a much longer-lived buffer than the rest of this snapshot (days, not
    /// the Resource trends chart's 1h) - see <c>HostStats.HostStatsPoller</c>'s remarks.
    /// <see langword="null"/> until <c>HostStatsOptions.DiskGrowthMinimumSpan</c> worth of
    /// samples has accumulated (a rate extrapolated from a few minutes of data is noise
    /// dressed up as a number) - the frontend omits the growth line entirely rather than
    /// showing a placeholder while this is <see langword="null"/>.
    /// </summary>
    public double? DiskGrowthBytesPerDay { get; init; }

    /// <summary>How much actual span <see cref="DiskGrowthBytesPerDay"/> was computed over - always shown alongside it (e.g. "based on the trailing 6h") so an early, still-warming-up buffer never silently implies a stable day-over-day figure. Settles at (and stays near) 24 once the buffer holds a full day.</summary>
    public double DiskGrowthWindowHours { get; init; }

    /// <summary>Best-effort - see <c>HostStats.ProcFileReader.ParseDiskStats</c>'s remarks on the whole-disk-vs-partition heuristic this sums across. 0.0 on the very first sample, same reasoning as <see cref="CpuUsagePercent"/>.</summary>
    public double DiskReadBytesPerSecond { get; init; }

    public double DiskWriteBytesPerSecond { get; init; }

    /// <summary>Bytes/sec, Rx + Tx combined across every non-loopback interface - what the Resource trends chart's "Network" line still plots. See <see cref="NetworkRxBytesPerSecond"/>/<see cref="NetworkTxBytesPerSecond"/> for the split the Host overview panel's compact live line shows instead.</summary>
    public double NetworkBytesPerSecond { get; init; }

    public double NetworkRxBytesPerSecond { get; init; }

    public double NetworkTxBytesPerSecond { get; init; }

    /// <summary>Packets/sec, Rx + Tx combined - the least specific of these fields (no split requested), one number rather than two.</summary>
    public double NetworkPacketsPerSecond { get; init; }

    /// <summary>Seconds since the host booted, from <c>/proc/uptime</c> - not this process's own uptime.</summary>
    public double UptimeSeconds { get; init; }

    /// <summary><see langword="null"/> until the first successful sample completes.</summary>
    public DateTimeOffset? UpdatedAt { get; init; }
}

File: src/dashboard/src/lib/memorypack/HostStatsSnapshot.ts
// Hand-written companion to MemoryPack's generated TypeScript classes - NOT itself
// generated. Mirrors `src/Flare.Api/Model/HostStatsSnapshot.cs`'s `HostStatsSnapshot`
// field-for-field, in declared order. Can't carry `[GenerateTypeScript]` itself: its
// `PerCoreUsagePercent` member is an `IReadOnlyList<double>` (see
// `PipelineServiceBreakdown.ts`'s header comment for why `IReadOnlyList<T>` alone blocks
// `[GenerateTypeScript]`) and it also has its own `DateTimeOffset? UpdatedAt` - see
// `$lib/memorypack/date-time-offset.ts`'s header comment.

import { MemoryPackWriter } from '$lib/generated/memorypack/MemoryPackWriter.js';
import { MemoryPackReader } from '$lib/generated/memorypack/MemoryPackReader.js';
import { readNullableDateTimeOffset, writeNullableDateTimeOffset } from '$lib/memorypack/date-time-offset';

export class HostStatsSnapshot {
	available: boolean;
	unavailableReason: string | null;
	cpuUsagePercent: number;
	cpuCoreCount: number;
	loadAverage1m: number;
	perCoreUsagePercent: number[] | null;
	memoryTotalBytes: bigint;
	memoryUsedBytes: bigint;
	memoryAvailableBytes: bigint;
	swapTotalBytes: bigint;
	swapUsedBytes: bigint;
	diskTotalBytes: bigint;
	diskUsedBytes: bigint;
	diskAvailableBytes: bigint;
	diskGrowthBytesPerDay: number | null;
	diskGrowthWindowHours: number;
	diskReadBytesPerSecond: number;
	diskWriteBytesPerSecond: number;
	networkBytesPerSecond: number;
	networkRxBytesPerSecond: number;
	networkTxBytesPerSecond: number;
	networkPacketsPerSecond: number;
	uptimeSeconds: number;
	updatedAt: Date | null;

	constructor() {
		this.available = false;
		this.unavailableReason = null;
		this.cpuUsagePercent = 0;
		this.cpuCoreCount = 0;
		this.loadAverage1m = 0;
		this.perCoreUsagePercent = null;
		this.memoryTotalBytes = 0n;
		this.memoryUsedBytes = 0n;
		this.memoryAvailableBytes = 0n;
		this.swapTotalBytes = 0n;
		this.swapUsedBytes = 0n;
		this.diskTotalBytes = 0n;
		this.diskUsedBytes = 0n;
		this.diskAvailableBytes = 0n;
		this.diskGrowthBytesPerDay = null;
		this.diskGrowthWindowHours = 0;
		this.diskReadBytesPerSecond = 0;
		this.diskWriteBytesPerSecond = 0;
		this.networkBytesPerSecond = 0;
		this.networkRxBytesPerSecond = 0;
		this.networkTxBytesPerSecond = 0;
		this.networkPacketsPerSecond = 0;
		this.uptimeSeconds = 0;
		this.updatedAt = null;
	}

	static serialize(value: HostStatsSnapshot | null): Uint8Array {
		const writer = MemoryPackWriter.getSharedInstance();
		this.serializeCore(writer, value);
		return writer.toArray();
	}

	static serializeCore(writer: MemoryPackWriter, value: HostStatsSnapshot | null): void {
		if (value == null) {
			writer.writeNullObjectHeader();
			return;
		}

		writer.writeObjectHeader(24);
		writer.writeBoolean(value.available);
		writer.writeString(value.unavailableReason);
		writer.writeFloat64(value.cpuUsagePercent);
		writer.writeInt32(value.cpuCoreCount);
		writer.writeFloat64(value.loadAverage1m);
		writer.writeArray(value.perCoreUsagePercent, (writer, x) => writer.writeFloat64(x));
		writer.writeInt64(value.memoryTotalBytes);
		writer.writeInt64(value.memoryUsedBytes);
		writer.writeInt64(value.memoryAvailableBytes);
		writer.writeInt64(value.swapTotalBytes);
		writer.writeInt64(value.swapUsedBytes);
		writer.writeInt64(value.diskTotalBytes);
		writer.writeInt64(value.diskUsedBytes);
		writer.writeInt64(value.diskAvailableBytes);
		writer.writeNullableFloat64(value.diskGrowthBytesPerDay);
		writer.writeFloat64(value.diskGrowthWindowHours);
		writer.writeFloat64(value.diskReadBytesPerSecond);
		writer.writeFloat64(value.diskWriteBytesPerSecond);
		writer.writeFloat64(value.networkBytesPerSecond);
		writer.writeFloat64(value.networkRxBytesPerSecond);
		writer.writeFloat64(value.networkTxBytesPerSecond);
		writer.writeFloat64(value.networkPacketsPerSecond);
		writer.writeFloat64(value.uptimeSeconds);
		writeNullableDateTimeOffset(writer, value.updatedAt);
	}

	static deserialize(buffer: ArrayBuffer): HostStatsSnapshot | null {
		return this.deserializeCore(new MemoryPackReader(buffer));
	}

	static deserializeCore(reader: MemoryPackReader): HostStatsSnapshot | null {
		const [ok, count] = reader.tryReadObjectHeader();
		if (!ok) {
			return null;
		}

		const value = new HostStatsSnapshot();
		if (count == 24) {
			value.available = reader.readBoolean();
			value.unavailableReason = reader.readString();
			value.cpuUsagePercent = reader.readFloat64();
			value.cpuCoreCount = reader.readInt32();
			value.loadAverage1m = reader.readFloat64();
			value.perCoreUsagePercent = reader.readArray((reader) => reader.readFloat64());
			value.memoryTotalBytes = reader.readInt64();
			value.memoryUsedBytes = reader.readInt64();
			value.memoryAvailableBytes = reader.readInt64();
			value.swapTotalBytes = reader.readInt64();
			value.swapUsedBytes = reader.readInt64();
			value.diskTotalBytes = reader.readInt64();
			value.diskUsedBytes = reader.readInt64();
			value.diskAvailableBytes = reader.readInt64();
			value.diskGrowthBytesPerDay = reader.readNullableFloat64();
			value.diskGrowthWindowHours = reader.readFloat64();
			value.diskReadBytesPerSecond = reader.readFloat64();
			value.diskWriteBytesPerSecond = reader.readFloat64();
			value.networkBytesPerSecond = reader.readFloat64();
			value.networkRxBytesPerSecond = reader.readFloat64();
			value.networkTxBytesPerSecond = reader.readFloat64();
			value.networkPacketsPerSecond = reader.readFloat64();
			value.uptimeSeconds = reader.readFloat64();
			value.updatedAt = readNullableDateTimeOffset(reader);
		} else if (count > 24) {
			throw new Error("Current object's property count is larger than type schema, can't deserialize about versioning.");
		} else {
			if (count == 0) return value;
			value.available = reader.readBoolean();
			if (count == 1) return value;
			value.unavailableReason = reader.readString();
			if (count == 2) return value;
			value.cpuUsagePercent = reader.readFloat64();
			if (count == 3) return value;
			value.cpuCoreCount = reader.readInt32();
			if (count == 4) return value;
			value.loadAverage1m = reader.readFloat64();
			if (count == 5) return value;
			value.perCoreUsagePercent = reader.readArray((reader) => reader.readFloat64());
			if (count == 6) return value;
			value.memoryTotalBytes = reader.readInt64();
			if (count == 7) return value;
			value.memoryUsedBytes = reader.readInt64();
			if (count == 8) return value;
			value.memoryAvailableBytes = reader.readInt64();
			if (count == 9) return value;
			value.swapTotalBytes = reader.readInt64();
			if (count == 10) return value;
			value.swapUsedBytes = reader.readInt64();
			if (count == 11) return value;
			value.diskTotalBytes = reader.readInt64();
			if (count == 12) return value;
			value.diskUsedBytes = reader.readInt64();
			if (count == 13) return value;
			value.diskAvailableBytes = reader.readInt64();
			if (count == 14) return value;
			value.diskGrowthBytesPerDay = reader.readNullableFloat64();
			if (count == 15) return value;
			value.diskGrowthWindowHours = reader.readFloat64();
			if (count == 16) return value;
			value.diskReadBytesPerSecond = reader.readFloat64();
			if (count == 17) return value;
			value.diskWriteBytesPerSecond = reader.readFloat64();
			if (count == 18) return value;
			value.networkBytesPerSecond = reader.readFloat64();
			if (count == 19) return value;
			value.networkRxBytesPerSecond = reader.readFloat64();
			if (count == 20) return value;
			value.networkTxBytesPerSecond = reader.readFloat64();
			if (count == 21) return value;
			value.networkPacketsPerSecond = reader.readFloat64();
			if (count == 22) return value;
			value.uptimeSeconds = reader.readFloat64();
			if (count == 23) return value;
			value.updatedAt = readNullableDateTimeOffset(reader);
			if (count == 24) return value;
		}
		return value;
	}
}
